pd/src/app.rs

```diff
@@ -335,29 +333,106 @@ impl App {
                 .unwrap()
                 .add_transaction(verified_transaction);
 
-            increment_counter!("node_transactions_total");
+            metrics::increment_counter!("node_transactions_total");
             Ok(())
         }
     }
 
-    fn end_block(&mut self, end: EndBlock) -> response::EndBlock {
-        let epoch = self
+    fn end_block(&mut self, end: EndBlock) -> impl Future<Output = Result<Response, BoxError>> {
+        // Clone a handle to the PendingBlock that we can move into the future
+        // below.  It's important not to retain a guard after aquiring a lock,
+        // since we'll be locking in an async context.
+        let pending_block = self
             .pending_block
-            .as_mut()
+            .as_ref()
             .expect("pending_block must be Some in EndBlock")
-            .lock()
-            .unwrap()
-            .set_height(end.height);
-
-        // TODO: if necessary, set the EndBlock response to add validators
-        // at the epoch boundary
-        if end.height.unsigned_abs() == epoch.start_height().value() {
-            // Epoch boundary -- add/remove validators if necessary
-            tracing::info!("new epoch");
-            increment_counter!("epoch");
+            .clone();
+
+        let height: u64 = end.height.try_into().expect("height should be nonnegative");
+        let epoch = pending_block.lock().unwrap().set_height(height);
+
+        let state = self.state.clone();
+        async move {
+            if epoch.end_height().value() == height {
+                tracing::info!(
+                    ?height,
+                    ?epoch,
+                    "last block of epoch, processing rate updates"
+                );
+                metrics::increment_counter!("epoch");
+
+                tracing::debug!("fetching rate data from state");
+                let current_base_rate = state.base_rate_data(epoch.index).await?;
+                let current_rates = state.rate_data(epoch.index).await?;
+                tracing::debug!(?current_base_rate);
+                for current_rate in &current_rates {
+                    tracing::debug!(?current_rate);
+                }
+
+                // here's where we should compute updated rates (moved from state code)
+                // TODO @ava insert calls here
+
+                // move below into calls to methods in stake crate?
```

Comment:
@avahowell I moved your TODO along with moving the code out of the `State` impl; I filled in some really basic stub code below, but I guess the computations for computing new rates should move into somewhere in the `stake` crate (?)

Reply:
yes, I think it makes sense to have them in the stake crate. in a wip branch where i've implemented the rate calculations they're in `stake/src/rates.rs`